src/PureScript/CST/ModuleGraph.purs

```diff
@@ -0,0 +1,120 @@
+module PureScript.CST.ModuleGraph
+  ( moduleGraph
+  , sortModules
+  ) where
+
+import Prelude
+
+import Data.Array as Array
+import Data.Bifunctor (bimap)
+import Data.Either (Either(..))
+import Data.Foldable (all, foldl)
+import Data.List (List(..))
+import Data.List as List
+import Data.Map (Map)
+import Data.Map as Map
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
+import Data.Set (Set)
+import Data.Set as Set
+import Data.Tuple (Tuple(..))
+import Control.Alt (alt)
+import PureScript.CST.Types (ImportDecl(..), ModuleHeader(..), ModuleName, Name(..))
+
+type Graph a = Map a (Set a)
+
+moduleGraph :: forall e. Array (ModuleHeader e) -> Graph ModuleName
+moduleGraph = Map.fromFoldable <<< map go
+  where
+  go (ModuleHeader { name: Name { name }, imports }) =
+    Tuple name (Set.fromFoldable (map getImportName imports))
+
+  getImportName (ImportDecl { "module": Name { name } }) = name
+
+sortModules :: forall e. Array (ModuleHeader e) -> Either (Array (ModuleHeader e)) (Array (ModuleHeader e))
+sortModules moduleHeaders = do
+  let
+    knownModuleHeaders :: Map ModuleName (ModuleHeader e)
+    knownModuleHeaders =
+      moduleHeaders
+        # map (\header@(ModuleHeader { name: Name { name } }) -> Tuple name header)
+        # Map.fromFoldable
+
+    graph = moduleGraph moduleHeaders
+
+  bimap (Array.mapMaybe (flip Map.lookup knownModuleHeaders) <<< List.toUnfoldable) (Array.mapMaybe (flip Map.lookup knownModuleHeaders) <<< List.toUnfoldable) (topoSort graph)
+
+type TopoSortArgs a =
+  { roots :: Set a
+  , sorted :: List a
+  , usages :: Map a Int
+  }
+
+topoSort :: forall a. Ord a => Graph a -> Either (List a) (List a)
+topoSort graph = do
+  let mbResults = go { roots: startingModules, sorted: Nil, usages: importCounts }
+  map _.sorted mbResults
+  where
+  go
+    :: TopoSortArgs a
+    -> Either (List a) (TopoSortArgs a)
+  go { roots, sorted, usages } = case Set.findMin roots of
+    Nothing ->
+      if all (eq 0) usages then
+        Right { roots, sorted, usages }
+      else do
+        let
+          nonLeaf = Set.fromFoldable do
+            Tuple a count <- Map.toUnfoldable usages :: Array (Tuple a Int)
+            if count /= 0 && Map.lookup a graph /= Nothing && Map.lookup a graph /= Just (Set.empty) then
+              [ a ]
+            else
+              []
```

Comment:
I think this is `Map.filterWithKey ?fn >>> Map.keys`

Reply:
You're right